imi/src/main/java/org/motechproject/nms/imi/component/SubscriptionCompleter.java

```diff
@@ -0,0 +1,76 @@
+package org.motechproject.nms.imi.component;
+
+import org.joda.time.DateTime;
+import org.motechproject.event.MotechEvent;
+import org.motechproject.event.listener.annotations.MotechListener;
+import org.motechproject.nms.imi.domain.CallDetailRecord;
+import org.motechproject.nms.imi.domain.CallRetry;
+import org.motechproject.nms.imi.repository.CallRetryDataService;
+import org.motechproject.nms.imi.service.RequestId;
+import org.motechproject.nms.kilkari.domain.Subscription;
+import org.motechproject.nms.kilkari.service.SubscriptionService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+/**
+ * Listens to nms.imi.complete_subscription MOTECH message and deactivates the provided subscription which,
+ * presumably, was rejected by IMI because the subscriber (from MCTS) is in the do not disturb (DND) database
+ */
+@Component
+public class SubscriptionCompleter {
+    private static final String COMPLETE_SUBSCRIPTION = "nms.imi.complete_subscription";
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(SubscriptionCompleter.class);
+
+    private SubscriptionService subscriptionService;
+    private CallRetryDataService callRetryDataService;
+
+
+    @Autowired
+    public SubscriptionCompleter(SubscriptionService subscriptionService,
+                                 CallRetryDataService callRetryDataService) {
+        this.subscriptionService = subscriptionService;
+        this.callRetryDataService = callRetryDataService;
+    }
+
+
+    @MotechListener(subjects = { COMPLETE_SUBSCRIPTION })
+    public void completeSubscription(MotechEvent event) {
```

Comment:
I think so.  I imagine we would define a dto that the IVR interface has to produce and publish.  Then all event handling that result in changes to subscriptions would live in the kilkari module.  If we change ivr providers the new module just needs to recreate and publish those events.